_posts/2021-03-19-forecasting-time-series-with-torch_4/forecasting_4.Rmd

```diff
@@ -0,0 +1,570 @@
+  })
+  
+  cat(sprintf("\nEpoch %d, validation: loss: %3.5f \n", epoch, mean(valid_loss)))
+}
+
+```
+
+    # Epoch 1, training: loss: 0.83752 
+    # Epoch 1, validation: loss: 0.83167
+
+    # Epoch 2, training: loss: 0.72803 
+    # Epoch 2, validation: loss: 0.80804 
+
+    # ...
+    # ...
+
+    # Epoch 99, training: loss: 0.10385 
+    # Epoch 99, validation: loss: 0.21259 
+
+    # Epoch 100, training: loss: 0.10396 
+    # Epoch 100, validation: loss: 0.20975 
+
+# Evaluation
+
+For visual inspection, we pick a few forecasts from the test set.
+
+```{r}
+net$eval()
+
+test_preds <- vector(mode = "list", length = length(test_dl))
+
+i <- 1
+
+vic_elec_test <- vic_elec_daily %>%
+  filter(year(Date) == 2014, month(Date) %in% 1:4)
+
+
+coro::loop(for (b in test_dl) {
+  
+  input <- b$x
+  output <- net(b$x$to(device = device), b$y$to(device = device), teacher_forcing_ratio = 0)
+  preds <- as.numeric(output)
+  
+  test_preds[[i]] <- preds
+  i <<- i + 1
+  
+})
+
+test_pred1 <- test_preds[[1]]
+test_pred1 <- c(rep(NA, n_timesteps), test_pred1, rep(NA, nrow(vic_elec_test) - n_timesteps - n_forecast))
+
+test_pred2 <- test_preds[[21]]
+test_pred2 <- c(rep(NA, n_timesteps + 20), test_pred2, rep(NA, nrow(vic_elec_test) - 20 - n_timesteps - n_forecast))
+
+test_pred3 <- test_preds[[41]]
+test_pred3 <- c(rep(NA, n_timesteps + 40), test_pred3, rep(NA, nrow(vic_elec_test) - 40 - n_timesteps - n_forecast))
+
+test_pred4 <- test_preds[[61]]
+test_pred4 <- c(rep(NA, n_timesteps + 60), test_pred4, rep(NA, nrow(vic_elec_test) - 60 - n_timesteps - n_forecast))
+
+test_pred5 <- test_preds[[81]]
+test_pred5 <- c(rep(NA, n_timesteps + 80), test_pred5, rep(NA, nrow(vic_elec_test) - 80 - n_timesteps - n_forecast))
+
+
+preds_ts <- vic_elec_test %>%
+  select(Demand, Date) %>%
+  add_column(
+    ex_1 = test_pred1 * train_sd + train_mean,
+    ex_2 = test_pred2 * train_sd + train_mean,
+    ex_3 = test_pred3 * train_sd + train_mean,
+    ex_4 = test_pred4 * train_sd + train_mean,
+    ex_5 = test_pred5 * train_sd + train_mean) %>%
+  pivot_longer(-Date) %>%
+  update_tsibble(key = name)
+
+
+preds_ts %>%
+  autoplot() +
+  scale_color_hue(h = c(80, 300), l = 70) +
+  theme_minimal()
+
+```
+
+```{r, eval = TRUE, echo = FALSE, layout="l-page", fig.cap="A sample of two-weeks-ahead predictions for the test set, 2014."}
+knitr::include_graphics("images/att_preds.png")
+```
+
+Of course, we now can't compare performance to that of previous models in our series, as we've pragmatically redefined the task. The main goal, however, anyway has been to introduce the concept of attention. Specifically, I wanted to show how to *manually* implement the technique -- something that, once you've understood the concept, you may never have to due in practice. Instead, you would likely make use of existing tools that come with `torch` (multi-head attention and transformer modules), tools we may introduce in a future "season" of this series.
```

Comment:
```suggestion
We can't directly compare performance here to that of previous models in our series, as we've pragmatically redefined the task. The main goal, however, has been to introduce the concept of attention. Specifically, how to *manually* implement the technique -- something that, once you've understood the concept, you may never have to due in practice. Instead, you would likely make use of existing tools that come with `torch` (multi-head attention and transformer modules), tools we may introduce in a future "season" of this series.
```

Reply:
again, removing the stuff I didn't like myself :-)))